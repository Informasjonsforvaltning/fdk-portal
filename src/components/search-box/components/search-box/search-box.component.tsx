import React, {
  ChangeEvent,
  Children,
  FC,
  isValidElement,
  memo,
  PropsWithChildren,
  useEffect,
  useState
} from 'react';

import { compose } from 'redux';
import SC from './styled';
import SearchForm from '../search-form/search-form.component';
import SearchLink from '../search-link/search-link.component';
import SearchBoxHeader from '../search-box-header/search-box-header.component';
import { Tabs } from '../../../../pages/search-page/tabs/tabs';
import withSearchSuggestions, {
  Props as SearchSuggestionProps
} from '../../../with-suggestions';
import { getConfig } from '../../../../config';

const SearchBox: FC<PropsWithChildren<SearchSuggestionProps>> = ({
  children,
  cachedSuggestions,
  searchSuggestionsActions: {
    getSearchSuggestionsRequested: getSearchSuggestions,
    resetSearchSuggestions
  }
}) => {
  const isNap = getConfig().filterTransportDatasets;

  const [searchString, setSearchString] = useState('');

  const renderSearchBoxHeader = () =>
    Children.map(children, child =>
      isValidElement(child) && child.type === SearchBoxHeader ? child : null
    )?.shift();

  const renderSearchLinks = () =>
    Children.map(children, child =>
      isValidElement(child) && child.type === SearchLink ? child : null
    );

  const renderTabs = () =>
    Children.map(children, child =>
      isValidElement(child) && child.type === Tabs ? child : null
    )?.shift();

  const searchChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearchString(e.target.value);
    if (!cachedSuggestions[e.target.value] && e.target.value.length > 1) {
      getSearchSuggestions(e.target.value, isNap);
    }
  };

  useEffect(
    () => () => {
      resetSearchSuggestions();
    },
    []
  );

  return (
    <SC.SearchBox>
      <SC.Content>
        {renderSearchBoxHeader()}
        <SearchForm
          suggestions={cachedSuggestions[searchString]}
          onChangeHandler={searchChange}
        />
        {Children.count(renderSearchLinks()) > 0 && (
          <SC.SearchLinks>{renderSearchLinks()}</SC.SearchLinks>
        )}
      </SC.Content>
      {renderTabs()}
    </SC.SearchBox>
  );
};

export default compose<FC<any>>(withSearchSuggestions, memo)(SearchBox);
