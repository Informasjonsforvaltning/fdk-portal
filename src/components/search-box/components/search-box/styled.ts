import styled, { css } from 'styled-components';
import { Colour, Unit, theme } from '@fellesdatakatalog/theme';

import SearchBoxHeaderSC from '../search-box-header/styled';
import SearchBoxBackground from '../../../../img/illustration-search-dark.svg';
import { getConfig } from '../../../../config';

const isTransportPortal = getConfig().themeNap;

const SearchBox = styled.section`
  align-items: center;
  background-repeat: no-repeat;
  background-position: top right;
  background-image: url(${SearchBoxBackground});
  color: ${theme.colour(Colour.NEUTRAL, 'N0')};
  display: flex;
  flex-flow: column;
  font-size: ${theme.fontSize('FS16', Unit.REM)};
  margin-bottom: 1em;

  ${({ theme: extendedTheme }) =>
    isTransportPortal
      ? css`
          background-color: ${extendedTheme.extendedColors.neutralDarkest};
        `
      : css`
          background-color: ${theme.colour(Colour.NEUTRAL, 'N60')};
        `}

  @media (max-width: 900px) {
    background-size: 40%;
    align-items: stretch;
  }
`;

const Content = styled.div`
  align-items: center;
  display: flex;
  flex: 1 1 auto;
  flex-flow: column;
  justify-content: space-between;
  padding: 0 15px;
  margin-bottom: 4em;
  margin-top: 4em;

  @media (min-width: 768px) {
    max-width: 720px;
    width: 100%;
  }

  @media (min-width: 992px) {
    max-width: 960px;
    width: 100%;
  }

  @media (min-width: 1200px) {
    max-width: 1140px;
    width: 100%;
  }

  ${SearchBoxHeaderSC.SearchBoxHeader} {
    margin-bottom: 1em;
    text-align: center;
  }
`;

const SearchLinks = styled.div`
  display: flex;
  flex-flow: column;
  justify-content: space-between;
  margin-top: 4em;
  width: 100%;

  @media (min-width: 992px) {
    flex-flow: row;
  }
`;

const SearchHeaderLogosTitle = styled.div`
  font-size: ${theme.fontSize('FS16', Unit.REM)};
  font-weight: 400;
`;

export default {
  SearchBox,
  Content,
  SearchLinks,
  SearchHeaderLogosTitle
};
