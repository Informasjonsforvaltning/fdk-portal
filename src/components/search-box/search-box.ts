export { default as SearchBox } from './components/search-box/search-box.component';

export { default as SearchBoxTitle } from './components/search-box-title/search-box-title.component';
