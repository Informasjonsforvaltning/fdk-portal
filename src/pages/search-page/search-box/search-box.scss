@import '../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

/*  Search box clear-icon for mobile devices */
.theme-fdk {
  .clear-icon:after {
    background-image: url('../../../img/icon-clear.svg');
  }
}

.theme-nap {
  .clear-icon:after {
    background-image: url('../../../img/icon-clear-nap.svg');
  }
}

.clear-icon:after {
  content: '';
  display: block;
  vertical-align: middle;
  outline: 0;
  cursor: pointer;
  position: absolute;
  z-index: 1;
  width: 0.7em;
  height: 0.7em;
  right: 15px;
  top: 20px;
}

.fdk-search-flex {
  display: flex;
  align-items: flex-start;
  font-size: 2rem;

  .fdk-search-box {
    position: relative;
    .glyphicon-search-frontpage {
      top: 20px;
    }
    label {
      width: 100%
    }
    input.fdk-search {
      @include themify($themes) {
        color: themed('color-neutral-darkest');
      }
      border: none;
      border-radius: 0;
      box-shadow: 0 4px 8px rgba(45, 55, 65, 0.2);
      font-weight: 400;
      height: 54px;
      width: 100%;
      padding: 12px 16px;
      margin: 0 0 10px 0;
    }
  }

  .fdk-button-filter {
    height: 54px;
    margin: 0;
    border-radius: 5px 0 0 5px;
    box-shadow: 0 4px 8px rgba(45, 55, 65, 0.2);
    border: 0;
    @include themify($themes) {
      background-color: themed('color-button-primary-1');
      color: themed('color-button-text-primary');
      border-color: themed('color-button-primary-1');
    }
    &:hover, &:focus {
      @include themify($themes) {
        color: themed('color-button-text-primary');
        border-color: themed('color-button-primary-1');
      }
    }
  }

  .fdk-button-search {
    margin: 0;
    border-radius: 0 5px 5px 0;
    box-shadow: 0 4px 8px rgba(45, 55, 65, 0.2);
    height: 54px;
    @include themify($themes) {
      background-color: themed('color-button-primary-1');
      color: themed('color-button-text-primary');
      border-color: themed('color-button-primary-1');
    }
    &:hover, &:focus {
      @include themify($themes) {
        color: themed('color-button-text-primary');
        border-color: themed('color-button-primary-1');
      }
    }
  }
}

/* Medium Devices, Desktops */
@media only screen and (min-width : 992px) {

  /*  Search box clear-icon for desktops */
  .clear-icon:after {
    width: 1em;
    height: 1em;
    right: 20px;
    top: 24px;
  }

  .fdk-search-flex {
    .fdk-search-box {
      input.fdk-search {
        border-radius: 5px 0 0 5px;
        font-size: 3rem;
        line-height: 68px;
        height: 68px;
      }
    }

    .fdk-button-search {
      font-size: 2.5rem;
      height: 68px;
    }
  }

  .glyphicon-search-frontpage {
    border-radius: 5px;
    right: 26px ;
    top: 92px;
    font-size: 3rem;
  }

}

/* Large Devices, Wide Screens */
@media only screen and (min-width : 1200px) {

}
