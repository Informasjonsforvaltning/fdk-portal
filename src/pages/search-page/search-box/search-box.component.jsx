import React from 'react';
import PropTypes from 'prop-types';
import { withState, withHandlers, compose } from 'recompose';
import _ from 'lodash';

import { getParamFromUrl } from '../../../lib/addOrReplaceUrlParam';
import localization from '../../../lib/localization';
import { HitsStats } from './hits-stats/hits-stats.component';
import './search-box.scss';

export const SearchBoxPure = props => {
  const {
    onSearchSubmit,
    countDatasets,
    countTerms,
    countApis,
    countInformationModels,
    open,
    searchText,
    inputText,
    setInputText,
    clearInputTextHandler,
    touched
  } = props;
  let refSearchBox; // eslint-disable-line no-unused-vars
  return (
    <div className="container pt-5 pb-5">
      <div className="col-12 col-lg-10 offset-lg-1 fdk-search-flex d-flex">
        <div className="d-inline d-lg-none">
          <button
            type="button"
            className="fdk-button fdk-button-filter"
            onClick={open}
          >
            Filter
          </button>
        </div>
        <div className="fdk-search-box flex-grow-1">
          <form
            onSubmit={e => {
              e.preventDefault();
              onSearchSubmit(inputText);
            }}
          >
            <label htmlFor="searchBox">
              <input
                name="searchBox"
                ref={input => {
                  refSearchBox = input;
                }}
                type="search"
                placeholder={localization.query.intro}
                aria-label={localization.query.intro}
                className="fdk-search"
                value={touched ? inputText : searchText}
                onChange={e => setInputText(e)}
                autoComplete="off"
              />
              <div
                role="button"
                tabIndex="0"
                className="clear-icon"
                aria-label={localization.query.reset}
                onClick={e => {
                  clearInputTextHandler(e);
                }}
                onKeyDown={() => {}}
              />
            </label>
          </form>
        </div>
        <button
          type="button"
          onClick={e => {
            e.preventDefault();
            onSearchSubmit(inputText);
          }}
          className="fdk-button-search btn btn-lg"
        >
          <i className="fa fa-search mr-2" />
          {localization.query.do}
        </button>
      </div>
      <div className="col-md-12 text-center">
        <HitsStats
          countDatasets={countDatasets}
          countTerms={countTerms}
          countApis={countApis}
          countInformationModels={countInformationModels}
          filteringOrTextSearchPerformed={
            !!(
              getParamFromUrl('q') ||
              getParamFromUrl('theme') ||
              getParamFromUrl('accessRight') ||
              getParamFromUrl('publisher')
            )
          }
        />
      </div>
    </div>
  );
};

SearchBoxPure.defaultProps = {
  searchText: null,
  countDatasets: null,
  countTerms: null,
  countApis: null,
  countInformationModels: null,
  inputText: null,
  setInputText: _.noop,
  clearInputTextHandler: _.noop,
  touched: false
};

SearchBoxPure.propTypes = {
  onSearchSubmit: PropTypes.func.isRequired,
  searchText: PropTypes.string,
  countDatasets: PropTypes.number,
  countTerms: PropTypes.number,
  countApis: PropTypes.number,
  countInformationModels: PropTypes.number,
  open: PropTypes.func.isRequired,
  inputText: PropTypes.string,
  setInputText: PropTypes.func,
  clearInputTextHandler: PropTypes.func,
  touched: PropTypes.bool
};

const enhance = compose(
  withState('inputText', 'setInputText', ''),
  withState('touched', 'setTouched', false),
  withHandlers({
    setInputText: props => event => {
      event.preventDefault();
      props.setTouched(true);
      props.setInputText(event.target.value);
    },
    clearInputTextHandler: props => event => {
      event.preventDefault();
      props.setTouched(false);
      props.setInputText('');
      props.onSearchSubmit('');
    }
  })
);

export const SearchBoxWithState = enhance(SearchBoxPure);
export const SearchBox = SearchBoxWithState;
