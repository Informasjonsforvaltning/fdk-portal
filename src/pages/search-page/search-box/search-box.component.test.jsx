import React from 'react';
import { shallow } from 'enzyme';
import { SearchBoxPure } from './search-box.component';

let onSearchSubmit;
let setInputText;
let preventDefault;
let open;
let defaultProps;

beforeEach(() => {
  onSearchSubmit = jest.fn();
  setInputText = jest.fn();
  open = jest.fn();
  preventDefault = jest.fn();
  defaultProps = {
    onSearchSubmit,
    setInputText,
    inputText: 'queryToSubmit',
    searchText: 'enhetsregister',
    countDatasets: 100,
    countTerms: 10,
    open
  };
});

test('should render SearchBoxPure correctly', () => {
  const wrapper = shallow(<SearchBoxPure {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should call onSearchSubmit prop for valid form submission', () => {
  const mockedEvent = {
    target: { value: 'test value' },
    preventDefault
  };
  const wrapper = shallow(<SearchBoxPure {...defaultProps} />);
  wrapper.find('form').prop('onSubmit')(mockedEvent);
  expect(preventDefault).toHaveBeenCalled();
  expect(onSearchSubmit).toHaveBeenLastCalledWith('queryToSubmit');
});

test('should call setInputText on change', () => {
  const mockedEvent = {
    target: { value: 'test value' },
    preventDefault
  };
  const wrapper = shallow(<SearchBoxPure {...defaultProps} />);
  wrapper.find('input').prop('onChange')(mockedEvent);
  expect(setInputText).toHaveBeenLastCalledWith(mockedEvent);
});

test('should call onSearchSubmit prop for valid form submission', () => {
  const mockedEvent = {
    target: { value: 'test value' },
    preventDefault
  };
  const wrapper = shallow(<SearchBoxPure {...defaultProps} />);
  wrapper.find('button.fdk-button-search').prop('onClick')(mockedEvent);
  expect(onSearchSubmit).toHaveBeenLastCalledWith('queryToSubmit');
});

test('should call open when filter clicked', () => {
  const wrapper = shallow(<SearchBoxPure {...defaultProps} />);
  wrapper.find('button.fdk-button-filter').prop('onClick')();
  expect(open).toHaveBeenCalled();
});
