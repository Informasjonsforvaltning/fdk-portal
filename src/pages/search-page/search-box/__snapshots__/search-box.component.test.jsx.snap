// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchBoxPure correctly 1`] = `
<div
  className="container pt-5 pb-5"
>
  <div
    className="col-12 col-lg-10 offset-lg-1 fdk-search-flex d-flex"
  >
    <div
      className="d-inline d-lg-none"
    >
      <button
        className="fdk-button fdk-button-filter"
        onClick={[MockFunction]}
        type="button"
      >
        Filter
      </button>
    </div>
    <div
      className="fdk-search-box flex-grow-1"
    >
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="searchBox"
        >
          <input
            aria-label="Search for contents"
            autoComplete="off"
            className="fdk-search"
            name="searchBox"
            onChange={[Function]}
            placeholder="Search for contents"
            type="search"
            value="enhetsregister"
          />
          <div
            aria-label="Clear"
            className="clear-icon"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex="0"
          />
        </label>
      </form>
    </div>
    <button
      className="fdk-button-search btn btn-lg"
      onClick={[Function]}
      type="button"
    >
      <i
        className="fa fa-search mr-2"
      />
      Search
    </button>
  </div>
  <div
    className="col-md-12 text-center"
  >
    <HitsStats
      countApis={null}
      countDatasets={100}
      countInformationModels={null}
      countTerms={10}
      filteringOrTextSearchPerformed={false}
    />
  </div>
</div>
`;
